### PR TITLE
chore: integrate Keycloak

### PR DESCRIPTION
**Pull request by MaxSchaefer**
## Which issues does this pull request close?
closes #

### CODE REVIEW

**Comment by MaxSchaefer** [comment]
Blocked by https://github.com/helpwave/services/pull/832

File: tasks/pages/templates.tsx [hunk 1]
```diff
@@ -83,6 +83,10 @@ const PersonalTaskTemplatesPage: NextPage = ({ overwriteTranslation }: PropsForT
   const { user } = useAuth()
   const { isLoading, isError, data } = usePersonalTaskTemplateQuery(user?.id)
 
+  useEffect(() => {
+    console.log('PTTP', user) //
+  }, [user])
+
```
Comment (DasProffi) [inline tasks/pages/templates.tsx:89]: Remove the logging here, or does it serve some purpose for production usage?

File: tasks/utils/grpc.ts [hunk 2]
```diff
@@ -31,26 +26,16 @@ export const organizationService = new OrganizationServicePromiseClient(userSvcB
...
-
-  if (organization) {
-    organizationID = organization.id
-  } else {
-    console.warn('Fallback to default organization')
-  }
+export const getAuthenticatedGrpcMetadata = (_: string = defaultOrganization): AuthenticatedGrpcMetadata => {
+  const token = getCurrentTokenAndUpdateInBackground()
+  return { Authorization: `Bearer ${token}` }
+}
```
Comment (DasProffi) [inline tasks/utils/grpc.ts:36]: Move this method and the config for the authentication to the api package as done in https://github.com/helpwave/web/pull/1083

**Comment by MaxSchaefer** [comment]
@DasProffi https://github.com/helpwave/services/pull/832 is merged

File: api-services/offline/tasks/bed_service.ts [hunk 3]
```diff
@@ -144,25 +139,6 @@ export class BedOfflineServicePromiseClient extends BedServicePromiseClient {
...
-      BedOfflineService.addBed(bed)
-    }
-
-    const list = beds.map(bed => new BulkCreateBedsResponse.Bed()
-      .setId(bed.id)
-      .setName(bed.name))
-
-    return new BulkCreateBedsResponse().setBedsList(list)
-  }
-
```
Comment (DasProffi) [inline api-services/offline/tasks/bed_service.ts:165]: Why did you remove this? The method still exists on the API

File: api-services/offline/users/organization_service.tsx [hunk 4]
```diff
@@ -153,22 +158,6 @@ export class OrganizationOfflineServicePromiseClient extends OrganizationService
...
-      email: request.getContactEmail(),
-      isPersonal: request.getIsPersonal(),
-      isVerified: true
-    }
-
-    OrganizationOfflineService.create(newOrganization)
-
-    return new CreateOrganizationForUserResponse().setId(newOrganization.id)
-  }
-
```
Comment (DasProffi) [inline api-services/offline/users/organization_service.tsx:171]: Removing this will have the effect that super is called and the GRPC client makes a backend request. For correct offline behavior, this method should stay implemented

File: tasks/components/layout/PageWithHeader.tsx [hunk 5]
```diff
@@ -28,14 +24,12 @@ export const PageWithHeader = ({
...
-  const { user } = useAuth()
-  const [isOrganizationSwitchModalOpen, setOrganizationSwitchModalOpen] = useState(false)
-  const [organization, setOrganization] = useState<OrganizationDTO>()
+  const { user, redirectUserToOrganizationSelection } = useAuth()
 
   if (!user) return null
 
   const feedbackButton = <FeedbackButton/>
-  const organizationName = (organization?.shortName && <Span onClick={() => setOrganizationSwitchModalOpen(true)} className={tw('cursor-pointer hover:cursor-pointer')}>{organization.shortName}</Span>)
+  const organization = <div onClick={() => redirectUserToOrganizationSelection()}>{user?.organization?.name}</div>
```
Comment (DasProffi) [inline tasks/components/layout/PageWithHeader.tsx:32]: For correct tag semantic and the cursor-pointer effect on hover
```suggestion
  const organization = <button onClick={() => redirectUserToOrganizationSelection()}>{user?.organization?.name}</button>
```

File: tasks/package.json [hunk 6]
```diff
@@ -29,6 +29,7 @@
     "google-protobuf": "3.21.2",
     "grpc-web": "1.5.0",
     "js-cookie": "3.0.5",
+    "keycloak-js": "^25.0.5",
```
Comment (DasProffi) [inline tasks/package.json:32]: Tasks should not have to import keycloak

File: api-services/authentication/useAuth.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  id: obj.sub,
+  email: obj.email,
+  name: obj.name,
+  nickname: obj.preferred_username,
+  avatarUrl: `https://source.boringavatars.com/marble/80/${obj.sub}`,
+  organization: obj.organization
+}))
+
+export type User = z.output<typeof UserFromIdTokenClaims>
+
```
Comment (DasProffi) [inline api-services/authentication/useAuth.tsx:48]: Make the user object the same as for the user type in the users API or call it differently here for example identity

**Review by DasProffi — APPROVED** [review]

File: api-services/offline/tasks/bed_service.ts (hunk 3 above)
Comment (MaxSchaefer) [reply api-services/offline/tasks/bed_service.ts:165]: Removed since https://github.com/helpwave/services/commit/dcc524f1165b117d833803f8a874373440016cee#diff-2fa17908774eb944a99da99eeea50dcb5eb50a134388d74898306e2bac1534f2L20

Comment (MaxSchaefer) [reply api-services/offline/tasks/bed_service.ts:165]: cc. @FoseFx

File: api-services/offline/users/organization_service.tsx (hunk 4 above)
Comment (MaxSchaefer) [reply api-services/offline/users/organization_service.tsx:171]: Removed since https://github.com/helpwave/services/commit/12e8933752cf54502dbecfb72833f01060ccccdd#diff-5c30cc40fa91c587bf380a97a8624fd62d22fed92d8580f6b315c2cf2a45500cL34

File: api-services/authentication/useAuth.tsx (hunk 7 above)
Comment (MaxSchaefer) [reply api-services/authentication/useAuth.tsx:48]: I would stick with user for now. In service we never really introduced the concept of `Identity`. Its always a user. We called a user with a valid session since the beginning User. We should discuss this separately.

**Review by DasProffi — APPROVED** [review]
